Install Antora extensions in Antora container before running `antora`

## Issue

As quarkiverse uses kroki as tool to convert the images for the site, it would be great to add it too to this extension in order to build the documentation of an extension as we do on quarkiverse-docs to avoid such an error: `Cannot find module 'asciidoctor-kroki' at path /antora/docs/target `

```
2025-03-14 12:32:43,203 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkiverse.antora.deployment.AntoraProcessor#buildAntoraSite threw an exception: org.opentest4j.AssertionFailedError: null: Cannot find module 'asciidoctor-kroki' at path /antora/docs/target or /usr/local/share/.config/yarn/global/node_modules/@antora/asciidoc-loader/lib/config
Require stack:
- /usr/local/share/.config/yarn/global/node_modules/@antora/user-require-helper/lib/index.js
- /usr/local/share/.config/yarn/global/node_modules/@antora/cli/lib/cli.js
- /usr/local/share/.config/yarn/global/node_modules/@antora/cli/lib/index.js
- /usr/local/share/.config/yarn/global/node_modules/@antora/cli/bin/antora Add the --stacktrace option to see the cause of the error.
        at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:38)
        at org.junit.jupiter.api.Assertions.fail(Assertions.java:138)
        at io.quarkiverse.antora.deployment.NativeImageBuildRunner$AntoraFrameConsumer.assertNoErrors(NativeImageBuildRunner.java:218)
        at io.quarkiverse.antora.deployment.NativeImageBuildRunner$OutputSlurper.assertNoErrors(NativeImageBuildRunner.java:157)
        at io.quarkiverse.antora.deployment.NativeImageBuildRunner.build(NativeImageBuildRunner.java:109)
        at io.quarkiverse.antora.deployment.AntoraProcessor.buildWithContainer(AntoraProcessor.java:150)
        at io.quarkiverse.antora.deployment.AntoraProcessor.buildAntoraSite(AntoraProcessor.java:101)
        at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
        at java.base/java.lang.Thread.run(Thread.java:1583)
        at org.jboss.threads.JBossThread.run(JBossThread.java:499)

        at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:372)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:289)
        at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:61)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:89)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:428)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:107)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:70)

```

--- a/.gitignore
+++ b/.gitignore
@@ -62,3 +62,12 @@ pom.xml.tag
 pom.xml.releaseBackup
 pom.xml.versionsBackup
 release.properties
+
+# npm
+node_modules/
+package.json
+package-lock.json
+.npm/
+
+# Antora
+antora-cache/**

--- a/deployment/src/main/java/io/quarkiverse/antora/deployment/AntoraProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/antora/deployment/AntoraProcessor.java
@@ -10,6 +10,7 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiConsumer;
+import java.util.function.Consumer;
 import java.util.stream.Stream;
 
 import org.jboss.logging.Logger;
@@ -98,7 +99,7 @@ public class AntoraProcessor {
             }
         }
 
-        buildWithContainer(fixedConfig.image(), gitRepoRoot, antoraPlaybookPath);
+        buildWithContainer(fixedConfig.image(), gitRepoRoot, antoraPlaybookPath, pbInfo.npmPackages());
 
         try (Stream<Path> files = Files.walk(pbInfo.outDir)) {
             files.forEach(absP -> {
@@ -145,9 +146,10 @@ public class AntoraProcessor {
 
     }
 
-    private void buildWithContainer(String antoraImageName, final Path gitRepoRoot, final Path antoraPlaybookPath) {
+    private void buildWithContainer(String antoraImageName, final Path gitRepoRoot, final Path antoraPlaybookPath,
+            List<String> npmPackages) {
         try {
-            new NativeImageBuildRunner().build(antoraImageName, gitRepoRoot, antoraPlaybookPath);
+            new NativeImageBuildRunner().build(antoraImageName, gitRepoRoot, antoraPlaybookPath, npmPackages);
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
         } catch (IOException e) {
@@ -216,6 +218,8 @@ public class AntoraProcessor {
         output.put("dir", "./" + outputDir.toString());
 
         handleSupplementalFiles(playbook, (ui, oldValue) -> ui.put("supplemental_files", "./." + oldValue));
+        final List<String> npmPackages = new ArrayList<>();
+        handleExtensions(playbook, npmPackages::add);
 
         try (Writer out = Files.newBufferedWriter(augmentedAntoraPlaybookYml, StandardCharsets.UTF_8)) {
             yaml.dump(playbook, out);
@@ -223,7 +227,17 @@ public class AntoraProcessor {
             throw new RuntimeException("Could not write " + augmentedAntoraPlaybookYml, e);
         }
 
-        return new PlaybookInfo(augmentedAntoraPlaybookYml, targetDir.resolve(outputDir));
+        return new PlaybookInfo(augmentedAntoraPlaybookYml, targetDir.resolve(outputDir), npmPackages);
+    }
+
+    static void handleExtensions(Map<String, Object> playbook, Consumer<String> extensionConsumer) {
+        final Object asciidoc = playbook.get("asciidoc");
+        if (asciidoc instanceof Map) {
+            final Object extensions = ((Map<String, Object>) asciidoc).get("extensions");
+            if (extensions instanceof List) {
+                ((List<String>) extensions).stream().forEach(extensionConsumer);
+            }
+        }
     }
 
     static void handleSupplementalFiles(Map<String, Object> playbook,
@@ -273,7 +287,7 @@ public class AntoraProcessor {
         return result;
     }
 
-    private record PlaybookInfo(Path playbookPath, Path outDir) {
+    private record PlaybookInfo(Path playbookPath, Path outDir, List<String> npmPackages) {
 
     }
 }

--- a/deployment/src/main/java/io/quarkiverse/antora/deployment/NativeImageBuildRunner.java
+++ b/deployment/src/main/java/io/quarkiverse/antora/deployment/NativeImageBuildRunner.java
@@ -43,7 +43,7 @@ public class NativeImageBuildRunner {
         containerName = "antora-" + RandomStringUtils.random(5, true, false);
     }
 
-    public void build(String antoraImageName, Path outputDir, Path antoraPlaybookPath)
+    public void build(String antoraImageName, Path outputDir, Path antoraPlaybookPath, List<String> npmPackages)
             throws InterruptedException, IOException {
 
         final List<String> cmd = new ArrayList<>();
@@ -88,15 +88,40 @@ public class NativeImageBuildRunner {
         cmd.add("--name");
         cmd.add(containerName);
 
+        cmd.add("--entrypoint");
+        cmd.add("/bin/sh");
+
         cmd.add(antoraImageName);
 
-        cmd.add("--cache-dir=./antora-cache");
-        cmd.add(antoraPlaybookPath.toString());
+        cmd.add("-c");
+
+        final StringBuilder startScript = new StringBuilder();
+
+        if (!npmPackages.isEmpty()) {
+
+            /*
+             * Avoid the following npm error:
+             * Your cache folder contains root-owned files, due to a bug in
+             * previous versions of npm which has since been addressed.
+             */
+            startScript.append("mkdir -p ./antora-cache/npm-cache && export npm_config_cache=./antora-cache/npm-cache && ");
+
+            startScript.append("npm i");
+
+            for (String pkg : npmPackages) {
+                startScript.append(' ').append(pkg);
+            }
+            startScript.append(" && ");
+        }
+        startScript.append("antora --cache-dir=./antora-cache ");
+        startScript.append(antoraPlaybookPath.toString());
+
+        cmd.add(startScript.toString());
 
         final String[] buildCommand = cmd.toArray(new String[0]);
 
         log.infof("Running Antora with %s:", containerRuntime.getExecutableName());
-        log.info(String.join(" ", buildCommand).replace("$", "\\$"));
+        log.info("'" + String.join("' '", buildCommand).replace("$", "\\$") + "'");
         final Process process = new ProcessBuilder(buildCommand)
                 .directory(outputDir.toFile())
                 .redirectErrorStream(true)

--- a/dev-mode-test/antora-playbook.yml
+++ b/dev-mode-test/antora-playbook.yml
@@ -11,3 +11,8 @@ ui:
   bundle:
     url: https://gitlab.com/antora/antora-ui-default/-/jobs/artifacts/HEAD/raw/build/ui-bundle.zip?job=bundle-stable
     snapshot: true
+asciidoc:
+  attributes:
+    kroki-fetch-diagram: true
+  extensions:
+    - asciidoctor-kroki

--- a/dev-mode-test/modules/ROOT/pages/index.adoc
+++ b/dev-mode-test/modules/ROOT/pages/index.adoc
@@ -4,3 +4,11 @@ Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor i
 Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
 Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
 Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
+
+== Kroki extension
+
+.PlantUML example
+[plantuml,alice-and-bob,svg,role=sequence]
+....
+alice -> bob
+....

--- a/dev-mode-test/src/test/java/io/quarkiverse/antora/test/AntoraDevModeTest.java
+++ b/dev-mode-test/src/test/java/io/quarkiverse/antora/test/AntoraDevModeTest.java
@@ -4,8 +4,10 @@ import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
 
 import org.assertj.core.api.Assertions;
 import org.awaitility.Awaitility;
@@ -46,6 +48,20 @@ public class AntoraDevModeTest {
                         });
                 response
                         .body(CoreMatchers.containsString("<h1 class=\"page\">Lorem ipsum</h1>"));
+            }
+
+            /* Make sure the SVG file for the UML diagram was generated */
+            final Path imagesDir = baseDir.resolve("target/classes/META-INF/antora/quarkus-antora/dev/_images");
+            Assertions.assertThat(imagesDir).isDirectory();
+            try (Stream<Path> files = Files.list(imagesDir)) {
+                final Optional<String> svgFile = files
+                        .map(Path::getFileName)
+                        .map(Path::toString)
+                        .filter(fn -> fn.startsWith("alice-and-bob-") && fn.endsWith(".svg"))
+                        .findFirst();
+                Assertions.assertThat(svgFile)
+                        .withFailMessage(() -> "Expected " + imagesDir.toString() + "/alice-and-bob-*.svg to exist")
+                        .isNotEmpty();
             }
 
             /* Make sure new.adoc does not exist yet */

--- a/docs/modules/ROOT/pages/index.adoc
+++ b/docs/modules/ROOT/pages/index.adoc
@@ -83,6 +83,33 @@ image::live-edit.png[optional attribute,optional attribute]
 [[test-warnings]]
 Once you are done with editing, hit `Q` or `CRTL+C` to exit the dev mode.
 
+== `antora-playbook.yml`
+
+The Antora site generation process is controlled by the `https://docs.antora.org/antora/latest/playbook/[antora-playbook.yml]` file.
+
+If there is no `antora-playbook.yml` file found side by side with `antora.yml`, then `quarkus-antora` will generate one
+and store it under `target/antora-playbook.yml`.
+
+In case there is an `antora-playbook.yml` file in your Maven or Gradle module,
+then `quarkus-antora` will attempt to respect it as far as possible.
+`quarkus-antora` will transform som file paths so that they match the changed location of the augmented `antora-playbook.yml` file
+and so that the site is generated under `target/classes/META-INF/antora`.
+
+[TIP]
+====
+You can always check the result of the transformation under `target/antora-playbook.yml`.
+====
+
+=== Antora extensions
+
+In case there are any Antora extensions, such as `https://github.com/asciidoctor/asciidoctor-kroki[asciidoctor-kroki]` declared in your `antora-playbook.yml`,
+then `quarkus-antora` will install them using `npm i <extension-1>[ <extension-2>[ ...]]` inside the Antora container before invoking `antora`.
+
+[WARNING]
+====
+Local extension scripts are not supported at this time.
+====
+
 == Configuration
 
 include::./includes/quarkus-antora.adoc[]

